test/gtest/uct/test_peer_failure.cc

```diff
@@ -280,32 +302,35 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure, purge_failed_peer,
     send_recv_am(0);
     send_recv_am(1);
 
-    const ucs_time_t loop_end_limit = ucs::get_deadline();
-    const size_t num_pend_sends     = 64ul;
-    const uint64_t send_data        = 0;
+    const size_t num_pend_sends = 64ul;
+    const size_t max_retry_num  = 3ul;
     std::vector<pending_send_request_t> reqs(num_pend_sends);
 
     {
         scoped_log_handler slh(wrap_errors_logger);
 
-        ucs_status_t status;
-        do {
-            status = uct_ep_am_short(ep0(), 0, 0, &send_data,
-                                     sizeof(send_data));
-        } while ((status == UCS_OK) && (ucs_get_time() < loop_end_limit));
+        fill_resources();
 
-        if (status == UCS_OK) {
-            UCS_TEST_SKIP_R("unable to fill the UCT resources");
-        } else if (status != UCS_ERR_NO_RESOURCE) {
-            UCS_TEST_ABORT("AM Short failed with " << ucs_status_string(status));
+        kill_receiver();
+
+        size_t retry_cnt;
+        for (retry_cnt = 0; retry_cnt < max_retry_num; retry_cnt++) {
+            ucs_status_t status = add_pending(ep0(), reqs[0]);
+            if (UCS_OK == status) {
+                break;
+            } else {
+                EXPECT_EQ(UCS_ERR_BUSY, status);
+                fill_resources();
+            }
         }
 
-        kill_receiver();
+        if (retry_cnt == max_retry_num) {
+            UCS_TEST_SKIP_R("unable to add pending requests after "
+                            "filling UCT resources");
+        }
 
-        for (size_t i = 0; i < num_pend_sends; i ++) {
-            reqs[i].ep       = ep0();
-            reqs[i].uct.func = pending_cb;
-            EXPECT_EQ(UCS_OK, uct_ep_pending_add(ep0(), &reqs[i].uct, 0));
+        for (size_t i = 1; i < num_pend_sends; i++) {
+            EXPECT_UCS_OK(add_pending(ep0(), reqs[i]));
```

Comment:
i'd just skip the test if can not add to pending without retries

Reply:
do you mean if add_pending returns BUSY even once?

Reply:
> do you mean if add_pending returns BUSY even once?

it could be returned only from the 1st `add_pending()`, since pending callback don't return UCS_OK, it always tries to fill resource again

Reply:
yes, just do not always expect `UCS_OK`, because:
- it will not disable this test for UD, the test will be skipped just when async progress ticks before adding to pending
- proposed change does not guarantee that this test will always be valid for UD, but imo adds unnecessary complexity